lib/semantics/resolve-names.cc

```diff
@@ -3212,10 +3302,37 @@ void ConstructVisitor::Post(const parser::DoConstruct &x) {
 }
 
 void ConstructVisitor::Post(const parser::ConcurrentControl &x) {
-  auto &name{std::get<parser::Name>(x.t)};
-  if (auto *symbol{DeclareConstructEntity(name)}) {
-    CheckIntegerType(*symbol);
+  const auto &name{std::get<parser::Name>(x.t)};
+  auto *prev{FindSymbol(name)};
+  if (prev) {
+    if (prev->owner().kind() == Scope::Kind::Forall ||
+        prev->owner() == currScope()) {
+      SayAlreadyDeclared(name, *prev);
+      return;
+    }
+    name.symbol = nullptr;
   }
+  auto &symbol{DeclareObjectEntity(name, {})};
+  if (symbol.GetType()) {
+    // type came from explicit type-spec
+  } else if (!prev) {
+    ApplyImplicitRules(symbol);
+  } else if (!prev->has<ObjectEntityDetails>() && !prev->has<EntityDetails>()) {
+    Say2(name, "Index name '%s' conflicts with existing identifier"_err_en_US,
+        *prev, "Previous declaration of '%s'"_en_US);
+    return;
+  } else {
+    if (auto *type{prev->GetType()}) {
+      symbol.SetType(*type);
+    }
+    if (const auto *details{prev->detailsIf<ObjectEntityDetails>()}) {
+      if (details->IsArray()) {
+        Say2(name, "Index variable '%s' is not scalar"_err_en_US, *prev,
```

Comment:
Will this be caught by the call to `CheckScalarIntegerType` below? If the check is needed here there should probably be a return after the error is reported.